### PR TITLE
Add step-level performance profiling to the trainer

### LINKED ISSUE #717: [RFC]: Add Timing Metrics and Performance Tracing to Training and Data Generation Pipelines
### Motivation.

Speculators currently has a well-structured metric logging system (`speculators.metrics` logger with TensorBoard/W&B/Trackio/MLflow backends) that tracks model quality metrics (loss, accuracy, per-position acceptance rates), but has no instrumentation for performance. Other important parts of the pipeline like response regeneration also does not record any meaningful performance metric. There are lots of ongoing work for preprocessing & regeneration pipeline hardening, multi-node training and they can all benefit from a better performance tracing system.

### Proposed Change.

Add a lightweight step timer to the trainer that breaks each training step into phases and emits per-window averages through the existing `speculators.metrics` logger (TensorBoard / W&B / Trackio / MLflow).
Metrics:

- `profile/fetch_ms` -- data loading + host-to-device transfer
- `profile/fwd_ms` -- model forward pass
- `profile/bwd_ms` -- loss.backward + grad clipping
- `profile/opt_ms` -- optimizer step + LR scheduler
- `profile/step_ms` -- total step wall-clock
- `profile/tokens_per_s` -- throughput
- `profile/fetch_frac` -- fraction of step spent waiting on data (fetch-bound indicator)

Similarly, add timing to the response regeneration pipeline (`scripts/response_regeneration/`) and offline data generation (`scripts/data_generation_offline.py`):

- Per-request latency (already partially exists in response regeneration via `time.time()`, but not aggregated)
- Batch/window throughput (requests/s, tokens/s)
- Queue/wait time when the vLLM server is saturated
- Safetensors write time for data generation

Or any other reasonably design.

### Any Other Things.

_No response_

### PR DESCRIPTION
**Pull request by omerap12**
## Purpose
Add step-level performance profiling to the trainer (resolves #717, trainer portion).
Emits 7 new metrics through the existing `speculators.metrics` logger:
`profile/fetch_ms`, `profile/fwd_ms`, `profile/bwd_ms`, `profile/opt_ms`, `profile/step_ms`, `profile/tokens_per_s`, `profile/fetch_frac`.
## Tests
Add unit tests, can be run with:
```bash
python3 -m pytest tests/unit/train/test_step_timer.py -v
```
## Checklist

I have filled in:

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

File: src/speculators/train/trainer.py [hunk 1]
```diff
@@ -329,28 +363,39 @@ def train_epoch(self, epoch: int):
...
                 else v
                 for k, v in batch.items()
             }
 
+            timer.mark("fetch")
             _draft_tokens, loss, metrics = self.model(
                 **gpu_batch, **self.config.train_call_kwargs
             )
 
+            timer.mark("fwd")
```
[the remark this answers is not in the document]
Comment (omerap12) [reply src/speculators/train/trainer.py]: Fixed. thanks.

File: src/speculators/train/trainer.py [hunk 2]
```diff
@@ -329,28 +363,39 @@ def train_epoch(self, epoch: int):
...
             self._optimizers_step()
 
             current_lrs = {
                 type(opt).__name__: opt.param_groups[0]["lr"] for opt in self.optimizers
             }
             self._schedulers_step()
+            timer.mark("opt")
+
+            num_tokens = (gpu_batch["document_ids"] != -1).sum().item()
+            profile = timer.profile(num_tokens)
```
[the remark this answers is not in the document]
Comment (omerap12) [reply src/speculators/train/trainer.py]: Fixed. thanks.

**Comment by omerap12** [comment]
@coderabbitai review

**Review by orestis-z — COMMENTED** [review]
**Bug: `TypeError` crash when `log_freq > 1`**

When a non-profiled step runs, `timer.now()` returns `None`, so `t_before_fetch = None`. On the next profiled step, `timer.mark_value("start", None)` stores `None` in `_marks["start"]`, and `profile()` hits `m["fetch"] - None` → `TypeError`.

This means every profiled step except the very first one crashes when `log_freq > 1`. It only works with the default `log_freq == 1`.

**Suggested fix** in `train_epoch`:
```python
t_before_fetch = timer.now() or time.perf_counter()
```

This falls back to an unsynchronized timestamp on non-profiled steps — cheap and avoids the crash. The fetch timing on the next profiled step will still be approximately correct.

**Missing test coverage** — consider adding a test that exercises the `log_freq > 1` path, e.g. simulating two consecutive steps where only the second is profiled, to verify `t_before_fetch` carries a valid float across disabled→enabled transitions.

File: src/speculators/train/trainer.py [hunk 3]
```diff
@@ -37,6 +38,50 @@
 root_logger = logging.getLogger("speculators")
 metric_logger = logging.getLogger("speculators.metrics")
 
+
+class StepTimer:
```
Comment (orestis-z) [inline src/speculators/train/trainer.py]: Nit: since this is an implementation detail of the trainer (not imported or used elsewhere), consider prefixing with `_` → `_StepTimer` to signal it's internal.

Also, a new `StepTimer` is allocated every step. Minor overhead, but you could avoid it with a single timer that resets its `_marks` dict each step.

File: src/speculators/train/trainer.py [hunk 4]
```diff
@@ -37,6 +38,50 @@
...
+    def __init__(self, enabled: bool):
+        self.enabled = enabled
+        self._marks: dict[str, float] = {}
+
+    def mark(self, name: str) -> None:
+        if self.enabled:
+            torch.cuda.synchronize()
```
Comment (orestis-z) [inline src/speculators/train/trainer.py:56]: Each `mark()` call does a `torch.cuda.synchronize()`, and there are 4 marks + 1 `now()` = **5 CUDA syncs per profiled step**. This is correct for accurate phase timing (without sync, `perf_counter` doesn't reflect GPU work completion), but it serializes the CUDA pipeline and can make profiled steps slower than unprofiled ones.

Worth a brief docstring or inline note so users aren't surprised by the overhead — especially since `log_freq=1` (the default) means *every* step pays the cost.

**Review by orestis-z — CHANGES_REQUESTED** [review]

**Review by orestis-z — CHANGES_REQUESTED** [review]
Thanks for the contribution @omerap12!

**Comment by omerap12** [comment]
Thanks for the review @orestis-z, I believe I adjusted to your comments :)

**Review by orestis-z — APPROVED** [review]
LGTM — all comments addressed, tests pass. Nice work on the disabled→enabled transition test.

Minor nit (non-blocking): the `step_ms == 0` zero-division guard (lines 79-80) is still untested. A quick test where all timestamps are equal would cover it:

```python
def test_zero_step_ms_returns_zero_throughput(mock_sync):
    timer = _StepTimer(enabled=True)
    timer.mark_value("start", 1.0)
    with patch("...time.perf_counter", return_value=1.0):
        timer.mark("fetch")
        timer.mark("fwd")
        timer.mark("bwd")
        timer.mark("opt")
    profile = timer.profile(num_tokens=4096)
    assert profile["tokens_per_s"] == 0.0
    assert profile["fetch_frac"] == 0.0
```

**Comment by omerap12** [comment]
> LGTM — all comments addressed, tests pass. Nice work on the disabled→enabled transition test.
> 
> Minor nit (non-blocking): the `step_ms == 0` zero-division guard (lines 79-80) is still untested. A quick test where all timestamps are equal would cover it:
> 
> ```python
> def test_zero_step_ms_returns_zero_throughput(mock_sync):
>     timer = _StepTimer(enabled=True)
>     timer.mark_value("start", 1.0)
>     with patch("...time.perf_counter", return_value=1.0):
>         timer.mark("fetch")
>         timer.mark("fwd")
>         timer.mark("bwd")
>         timer.mark("opt")
>     profile = timer.profile(num_tokens=4096)
>     assert profile["tokens_per_s"] == 0.0
>     assert profile["fetch_frac"] == 0.0
> ```

Added. :)

**Comment by omerap12** [comment]
fix lint.

**Comment by orestis-z** [comment]
needs a `make quality`
